Data cleaning

Meng6 opened the pull request:
Updated the code for data cleaning:
- Fill NA with 0 for the selected event phone features within each sensor's feature extraction script
- Fill NA with 0 for the selected event phone features (do not consider plugin for this step, e.g. AR) in data cleaning script
- Add drop highly correlated features in data cleaning script

Need to do before merging:
- Update docs
- Update testing results
- Update example workflow

statsof1 commented:
Thanks @Meng6 Can you check the tests, please? They are falling for activity recognition

JulioV commented:
Thanks @Meng6 Can you check the tests, please? They are falling for activity recognition

Meng6 commented:
Thank you @JulioV. Weiyu is working on the testing results currently. Will let you know once it is done.

Meng6 commented:
Hi @JulioV, the testing results, example workflow, and docs were updated. Could you please double check when you are free? Thanks!

JulioV commented on docs/analysis/data-cleaning.md:
```diff
@@ -0,0 +1,81 @@
+Data Cleaning
+=============
+
+This module is to clean the extracted sensor features before merging it with the target labels.
```
The goal of this module is to perform basic clean tasks on the behavioral features that RAPIDS computes. You might need to do further processing depending on your analysis objectives.

JulioV commented on docs/analysis/data-cleaning.md:
```diff
@@ -0,0 +1,81 @@
+
+|Key&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;            | Description |
+|----------------|-----------------------------------------------------------------------------------------------------------------------------------
+|`[COMPUTE]` | Set to `True` to clean sensor features for individual participants from the `RAPIDS` provider|
```
Set to `True` to execute the cleaning tasks described below. You can use the parameters of each task to tweak them or deactivate them

JulioV commented on docs/analysis/data-cleaning.md:
```diff
@@ -0,0 +1,81 @@
+|Key&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;            | Description |
+|----------------|-----------------------------------------------------------------------------------------------------------------------------------
+|`[COMPUTE]` | Set to `True` to clean sensor features for individual participants from the `RAPIDS` provider|
+|`[IMPUTE_SELECTED_EVENT_FEATURES]`     | Fill NA with 0 for the selected event features, see table below
```
Fill NAs with 0 only for event-based features, see table below

JulioV commented on docs/analysis/data-cleaning.md:
```diff
@@ -0,0 +1,81 @@
+
+|Parameters                             | Description                                                    |
+|-------------------------------------- |----------------------------------------------------------------|
+|`[COMPUTE]`                            | Set to `True` to fill NA with 0 for the selected event features
```
Set to `True` to fill NAs with 0 for phone event-based features

JulioV replied:
We need to include the features that we are imputing in this step (maybe an expandable section?)

Meng6 replied:
Hi @JulioV, I am not sure how to list the selected event-based features as we use [regular expressions](https://github.com/carissalow/rapids/blob/data_cleaning/src/features/all_cleaning_individual/rapids/main.R#L18) to match these features. How about adding the following descriptions after [line49](https://github.com/carissalow/rapids/blob/data_cleaning/docs/analysis/data-cleaning.md?plain=1#L49)?

```
The following phone event-based features are considered currently:
- Application foreground features: countevent, countepisode, minduration, maxduration, meanduration, sumduration.
- Battery features: all features.
- Calls features: count, distinctcontacts, sumduration, minduration, maxduration, meanduration, modeduration.
- Keyboard features: sessioncount, averagesessionlength, changeintextlengthlessthanminusone, changeintextlengthequaltominusone, changeintextlengthequaltoone, changeintextlengthmorethanone, maxtextlength, totalkeyboardtouches.
- Messages features: count, distinctcontacts.
- Screen features: sumduration, maxduration, minduration, avgduration, countepisode.
- WiFi features: all connected and visible features.

JulioV replied:
Yep, this works

JulioV commented on docs/analysis/data-cleaning.md:
```diff
@@ -0,0 +1,81 @@
+|Parameters                             | Description                                                    |
+|-------------------------------------- |----------------------------------------------------------------|
+|`[COMPUTE]`                            | Set to `True` to fill NA with 0 for the selected event features
+|`[MIN_DATA_YIELDED_MINUTES_TO_IMPUTE]` | Assume the selected event sensor is working when phone_data_yield_rapids_ratiovalidyieldedminutes > `[MIN_DATA_YIELDED_MINUTES_TO_IMPUTE]`. |
```
Any feature value in a time segment instance with phone data yield > `[MIN_DATA_YIELDED_MINUTES_TO_IMPUTE]` will be replaced with a zero. See below for an explanation.

JulioV commented on docs/analysis/data-cleaning.md:
```diff
@@ -0,0 +1,81 @@
+|-------------------------------------- |----------------------------------------------------------------|
+|`[COMPUTE]`                            | Set to `True` to drop highly correlated features
+|`[MIN_OVERLAP_FOR_CORR_THRESHOLD]`     | Minimum ratio of observations required per pair of columns (features) to be considered as a valid correlation. 
+|`[CORR_THRESHOLD]` | The absolute values of pair-wise correlations are calculated. If two variables have a valid correlation higher than `[CORR_THRESHOLD]`, we looks at the mean absolute correlation of each variable and removes the variable with the largest mean absolute correlation.
```
we look at the mean absolute correlation of each variable and remove the variable with the largest mean absolute correlation using the R package [INCLUDE PACKAGE NAME]

JulioV commented on docs/analysis/data-cleaning.md:
```diff
@@ -0,0 +1,81 @@
+|`[MIN_OVERLAP_FOR_CORR_THRESHOLD]`     | Minimum ratio of observations required per pair of columns (features) to be considered as a valid correlation. 
+|`[CORR_THRESHOLD]` | The absolute values of pair-wise correlations are calculated. If two variables have a valid correlation higher than `[CORR_THRESHOLD]`, we looks at the mean absolute correlation of each variable and removes the variable with the largest mean absolute correlation.
+
+Steps to clean sensor features for individual participants. It only considers the **phone sensors** currently.
```
Steps to clean sensor features for individual participants. Currently, only **phone features ** are supported.
Write

JulioV commented on docs/analysis/data-cleaning.md:
```diff
@@ -0,0 +1,81 @@
+
+Steps to clean sensor features for individual participants. It only considers the **phone sensors** currently.
+
+??? info "1. Fill NA with 0 for the selected event features."
```
??? info "1. Replace NAs with 0 for selected phone event-based features."
Write

JulioV commented on docs/analysis/data-cleaning.md:
```diff
@@ -0,0 +1,81 @@
+Steps to clean sensor features for individual participants. It only considers the **phone sensors** currently.
+
+??? info "1. Fill NA with 0 for the selected event features."
+    Some event features should be zero instead of NA. In this step, we fill those missing features with 0 when the `phone_data_yield_rapids_ratiovalidyieldedminutes` column is higher than the `[IMPUTE_SELECTED_EVENT_FEATURES][MIN_DATA_YIELDED_MINUTES_TO_IMPUTE]` parameter. Plugins such as Activity Recognition sensor are not considered. You can skip this step by setting `[IMPUTE_SELECTED_EVENT_FEA ...
```
Some NA values in event-based phone features should be zero. In this step, we fill those missing values with 0 when the `phone_data_yield_rapids_ratiovalidyieldedminutes` column is higher than the `[IMPUTE_SELECTED_EVENT_FEATURES][MIN_DATA_YIELDED_MINUTES_TO_IMPUTE]` parameter. AWARE Framework plugins such as Activity Recognition sensor are not considered because they can run independently from the main app. You can skip this step by setting `[IMPUTE_SELECTED_EVENT_FEATURES][COMPUTE]` to `False`.

JulioV commented on docs/analysis/data-cleaning.md:
```diff
@@ -0,0 +1,81 @@
+??? info "1. Fill NA with 0 for the selected event features."
+    Some event features should be zero instead of NA. In this step, we fill those missing features with 0 when the `phone_data_yield_rapids_ratiovalidyieldedminutes` column is higher than the `[IMPUTE_SELECTED_EVENT_FEATURES][MIN_DATA_YIELDED_MINUTES_TO_IMPUTE]` parameter. Plugins such as Activity Recognition sensor are not considered. You can skip this step by setting `[IMPUTE_SELECTED_EVENT_FEA ...
+    
+    Take phone calls sensor as an example. If there are no calls records during a time segment for a participant, then (1) the calls sensor was not working during that time segment; or (2) the calls sensor was working and the participant did not have any calls during that time segment. To differentiate these two situations, we assume the selected sensors are working when `phone_data_yield_rapids_ ...
```
Take phone calls sensor as an example for a hypothetical participant. If there are no calls records during a time segment, then either (1) the calls sensor was not working during that time segment; or (2) the calls sensor was working *and* the participant did not have any calls during that time segment. To differentiate these two situations, we assume these event-based sensors are working when `phone_data_yield_rapids_ratiovalidyieldedminutes > [MIN_DATA_YIELDED_MINUTES_TO_IMPUTE]`. In other words, when we have enough data from other sensors.

JulioV commented on docs/analysis/data-cleaning.md:
```diff
@@ -0,0 +1,81 @@
+    Extracted features might be not reliable if the sensor only works for a short period during a time segment. In this step, we discard rows when the `phone_data_yield_rapids_ratiovalidyieldedhours` column is less than the `[DATA_YIELDED_HOURS_RATIO_THRESHOLD]` parameter. We do not recommend you to skip this step, but you can do it by setting `[DATA_YIELDED_HOURS_RATIO_THRESHOLD]` to 0.
+
+??? info "3. Discard columns (features) with too many missing values."
+    In this step, we discard columns with missing value ratios higher than `[COLS_NAN_THRESHOLD]`. We do not recommend you to skip this step, but you can do it by setting `[COLS_NAN_THRESHOLD]` to 1.
```
In this step, we discard columns with missing value ratios higher than `[COLS_NAN_THRESHOLD]`. We do not recommend you to skip this step, but you can do it by setting `[COLS_NAN_THRESHOLD]` to 1. In other words we are discarding features that did not have values in enough time segments (e.g. days) to be considered *reliable*.

JulioV commented on docs/analysis/data-cleaning.md:
```diff
@@ -0,0 +1,81 @@
+    Take phone calls sensor as an example. If there are no calls records during a time segment for a participant, then (1) the calls sensor was not working during that time segment; or (2) the calls sensor was working and the participant did not have any calls during that time segment. To differentiate these two situations, we assume the selected sensors are working when `phone_data_yield_rapids_ ...
+
+??? info "2. Discard unreliable rows."
+    Extracted features might be not reliable if the sensor only works for a short period during a time segment. In this step, we discard rows when the `phone_data_yield_rapids_ratiovalidyieldedhours` column is less than the `[DATA_YIELDED_HOURS_RATIO_THRESHOLD]` parameter. We do not recommend you to skip this step, but you can do it by setting `[DATA_YIELDED_HOURS_RATIO_THRESHOLD]` to 0.
```
Extracted features might be not reliable if the sensor only works for a short period during a time segment. In this step, we discard rows when the `phone_data_yield_rapids_ratiovalidyieldedhours` column is less than the `[DATA_YIELDED_HOURS_RATIO_THRESHOLD]` parameter. We do not recommend you to skip this step, but you can do it by setting `[DATA_YIELDED_HOURS_RATIO_THRESHOLD]` to 0. In other words we are discarding time segments (e.g. days) that did not have enough data to be considered *reliable*.

JulioV commented on docs/analysis/data-cleaning.md:
```diff
@@ -0,0 +1,81 @@
+    In this step, we discard columns with zero variance. We do not recommend you to skip this step, but you can do it by setting `[COLS_VAR_THRESHOLD]` to `False`.
+
+??? info "5. Drop highly correlated features."
+    As highly correlated features might not bring additional information and will increase the complexity of our model, we drop them in this step. The absolute values of pair-wise correlations are calculated. It is regarded as valid only if the ratio of this pair of columns (features) are less than `[DROP_HIGHLY_CORRELATED_FEATURES][MIN_OVERLAP_FOR_CORR_THRESHOLD]`. If two variables have a valid  ...
```
As highly correlated features might not bring additional information and will increase the complexity of a model, we drop them in this step. The absolute values of pair-wise correlations are calculated. Each correlation vector between two variables is regarded as valid only if the ratio of valid value pairs (i.e. non NA pairs) is less than `[DROP_HIGHLY_CORRELATED_FEATURES][MIN_OVERLAP_FOR_CORR_THRESHOLD]`. If two variables have a correlation coefficient higher than `[DROP_HIGHLY_CORRELATED_FEATURES][CORR_THRESHOLD]`, we look at the mean absolute correlation of each variable and remove the variable with the largest mean absolute correlation. This step can be skipped by setting `[DROP_HIGHLY_CORRELATED_FEATURES][COMPUTE]` to `False`.

JulioV replied:
In this sentence, should it say less or more than? `the ratio of valid value pairs (i.e. non NA pairs) is less than`

Meng6 replied:
Yes, you are right. It should be `greater than or equal to`.

JulioV commented on docs/analysis/data-cleaning.md:
```diff
@@ -0,0 +1,81 @@
+    As highly correlated features might not bring additional information and will increase the complexity of our model, we drop them in this step. The absolute values of pair-wise correlations are calculated. It is regarded as valid only if the ratio of this pair of columns (features) are less than `[DROP_HIGHLY_CORRELATED_FEATURES][MIN_OVERLAP_FOR_CORR_THRESHOLD]`. If two variables have a valid  ...
+
+??? info "6. Discard rows with too many missing values."
+    In this step, we discard rows with missing value ratios higher than `[ROWS_NAN_THRESHOLD]`. We do not recommend you to skip this step, but you can do it by setting `[ROWS_NAN_THRESHOLD]` to 1.
```
In this step, we discard rows with missing value ratios higher than `[ROWS_NAN_THRESHOLD]`. We do not recommend you to skip this step, but you can do it by setting `[ROWS_NAN_THRESHOLD]` to 1. In other words we are discarding time segments (e.g. days) that did not have enough data to be considered reliable. This step is similar to step 2 except the ratio is computed based on NA values instead of a phone data yield threshold.

JulioV commented on docs/change-log.md:
```diff
@@ -1,6 +1,8 @@
 # Change Log
 ## v.1.7.0
 - Add tests for phone battery features
+- Fill NA with 0 for the selected event features: (1) each sensor's feature extraction script (2) data cleaning script
```
Replace NA with 0 for selected event-based features. Done in each feature extraction script and the data cleaning module

JulioV commented on example_profile/example_config.yaml:
```diff
@@ -534,6 +534,44 @@ HEATMAP_FEATURE_CORRELATION_MATRIX:
+        CORR_THRESHOLD: 0.95
+      SRC_SCRIPT: src/features/all_cleaning_individual/rapids/main.R
+
+ALL_CLEANING_OVERALL:
```
Did we explain the difference between individual & overall in the docs?

Meng6 replied:
I think so. We have two sections in the docs: 
(1) Clean sensor features for individual participants; 
(2) Clean sensor features for all participants.

The only difference between those two sections is the input file. I listed the "File Sequence" for both sections to denote different input files: https://github.com/carissalow/rapids/blob/data_cleaning/docs/analysis/data-cleaning.md?plain=1#L68

Hi @JulioV, is it ok? Do I need to add descriptions at the beginning? For example, line4:
The goal of this module is to perform basic clean tasks on the behavioral features that RAPIDS computes. You might need to do further processing depending on your analysis objectives. This module can clean sensor features for both individual participants and all participants. You need to set parameters within each section (e.g. `[ALL_CLEANING_INDIVIDUAL]` and `[ALL_CLEANING_OVERALL]`).

JulioV replied:
Yeah let's add this. Small correction, this module can clean features at the individual level and at the study level. And give an example of why this is important please

Meng6 replied:
Thanks @JulioV. How about this example:

This module can clean features at the individual level and at the study level. For example, if we have enough data for a participant, we can train a model for that participant. By setting parameters in `[ALL_CLEANING_INDIVIDUAL]` section, we can get cleaned data for that participant. Similarly, if we want to train a model for all participants, we can get the cleaned data for all participants by setting parameters in `[ALL_CLEANING_OVERALL]` section.

JulioV replied:
This module can clean features at the individual level and at the study level. If you are interested in creating individual models (using each participant's features independently of the others) use `ALL_CLEANING_INDIVIDUAL`. If you are interested in creating population models (using everyones data in the same model) use `ALL_CLEANING_OVERALL`

I think this is clearer. What do you think?

Meng6 replied:
Yes! Thank you!

JulioV commented on docs/change-log.md:
```diff
@@ -1,6 +1,8 @@
 ## v.1.7.0
 - Add tests for phone battery features
+- Fill NA with 0 for the selected event features: (1) each sensor's feature extraction script (2) data cleaning script
+- Refactor data cleaning module: update the structure and add dropping highly correlated features section
```
Add a data cleaning module to replace NAs with 0 in selected event-based features, discard unreliable rows and columns, and discard columns with zero variance.

Meng6 replied:
Hi @JulioV , we also drop highly correlated features. Shall we update it as follows?

- Add a data cleaning module to replace NAs with 0 in selected event-based features, discard unreliable rows and columns, discard columns with zero variance, and discard highly correlated columns.

Meng6 commented:
Hi @JulioV, we just found that there might be an issue when time segment is very small. I split the `[DATA_YIELDED_HOURS_RATIO_THRESHOLD]` parameter into two parameters:  `DATA_YIELD_UNIT` and `[DATA_YIELD_RATIO_THRESHOLD]`. I also updated the docs as you suggested. Could you double check the latest docs? If it is good, I can merge it back to the develop branch! Thank you!

JulioV commented:
> Hi @JulioV, we just found that there might be an issue when time segment is very small. I split the `[DATA_YIELDED_HOURS_RATIO_THRESHOLD]` parameter into two parameters: `DATA_YIELD_UNIT` and `[DATA_YIELD_RATIO_THRESHOLD]`. I also updated the docs as you suggested. Could you double check the latest docs? If it is good, I can merge it back to the develop branch! Thank you!


Let's cheange `DATA_YIELD_UNIT` to `DATA_YIELD_FEATURE` and possible values to `RATIO_VALID_YIELDED_HOURS` & `RATIO_VALID_YIELDED_MUNUTES` 


Other than this, it's ready
